### PR TITLE
Adding NonNull type for validation of potentially null values.

### PR DESCRIPTION
**Pull request by ehoner**
This is a duplicate of fthomas/refined/issues/450, which is a request to provide safe `null` handling.

There is a discussion on the previous PR that I would like to address. I agree with @jan0sch [comment](https://github.com/fthomas/refined/issues/450#issuecomment-375655000) that it is the responsibility of the developer to handle `null`, but it seems reasonable that the library should afford the developer a mechanism to do so. I believe this patch would not cause any breaking changes and the maintenance overhead should be minimal.

### CODE REVIEW

File: modules/core/shared/src/test/scala/eu/timepit/refined/ImpureValidateSpec.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import eu.timepit.refined.impure.NonNull
+import eu.timepit.refined.numeric._
+import eu.timepit.refined.string._
+import org.scalacheck.Prop._
+import org.scalacheck.Properties
+
+class ImpureValidateSpec extends Properties("ImpureValidate") {
+
+  property("NonNull safe for String") = secure {
+    import eu.timepit.refined.collection.NonEmpty
```
Comment (Michaelt293) [inline modules/core/shared/src/test/scala/eu/timepit/refined/ImpureValidateSpec.scala]: I believe `NonEmpty` is already being imported on line 5 -`import eu.timepit.refined.collection._`?

Comment (ehoner) [reply modules/core/shared/src/test/scala/eu/timepit/refined/ImpureValidateSpec.scala]: You are correct.

**Comment by jan0sch** [comment]
Very nice! Does this work for all types which could be `null` or are there limits to it?

**Comment by ehoner** [comment]
@jan0sch 

I am not sure that I can say *all*, but the tests verify basic types `String` and `List` are safe when  `null`. With an additional test for `Positive` to show compatibility with the library - since `int` types cannot be null. There is also an example, with `URL`, where the `URL` type can be refined with the library (even if the value is `null`), but a valid `Refined[URL]` instance may return a `null` with `getUserInfo` and this PR can handle that safely. 

The goal was to make something specific to `null` handling that was compatible with the rest of the library. FWIW, all the tests that were immediately obvious to me have been included. I stumbled into this when I encountered the problem myself. Another demonstration based on the [comment](https://github.com/fthomas/refined/issues/450#issuecomment-387854403) in another PR, where the `Not` case was previously unsafe.

```scala
scala> val s: String = null
s: String = null

scala> refineV[Regex](s)
res3: Either[String,eu.timepit.refined.api.Refined[String,eu.timepit.refined.predicates.all.Regex]] = Left(Regex predicate failed: null)

scala> refineV[NonNull[Not[Regex]]](s)
res4: Either[String,eu.timepit.refined.api.Refined[String,eu.timepit.refined.predicates.all.NonNull[eu.timepit.refined.predicates.all.Not[eu.timepit.refined.predicates.all.Regex]]]] = Left(Predicate failed: source is null.)
```

**Comment by ehoner** [comment]
@fthomas do you have any thoughts?

**Comment by fthomas** [comment]
Sorry for the late reply, @ehoner.

I'm still wondering what the advantage of this predicate would be over composing `refineV` with a function that handles `null`s:
```scala
scala> def nonnull[T](t: T): Either[String, T] = if (t == null) Left("parameter is null") else Right(t)

scala> nonnull(null: String).flatMap(refineV[NonEmpty](_))
res1: scala.util.Either[String,eu.timepit.refined.api.Refined[String,eu.timepit.refined.predicates.all.NonEmpty]] = Left(parameter is null)

scala> nonnull("").flatMap(refineV[NonEmpty](_))
res2: scala.util.Either[String,eu.timepit.refined.api.Refined[String,eu.timepit.refined.predicates.all.NonEmpty]] = Left(Predicate isEmpty() did not fail.)

scala> nonnull("Hello!").flatMap(refineV[NonEmpty](_))
res3: scala.util.Either[String,eu.timepit.refined.api.Refined[String,eu.timepit.refined.predicates.all.NonEmpty]] = Right(Hello!)
```

**Comment by ehoner** [comment]
@fthomas 

I am not sure that I can definitively answer your question: benefits of function `nonnull` vs type `NonNull`. I can only offer that a type might better identify what refinement has taken place.

From my own experience, I created the PR when I first started using the library. I encountered the `null` problem immediately and my expectation was that the library would provide some mechanism for safely handling this. I didn't find any documentation/examples specific to `null`, but I did find a previous [issue](/issues/450). So I added what I thought was a way to handle `null` that is consistent with `refined`. This addition might be overkill, in fact I can't think of an example why the `NonNull` type would be strictly "necessary".

Baring acceptance of this PR, it might be valuable, for one of the examples to use the function above to help newcomers to the library. Or even add the function to the library to newcomers or provide a "refined" way to handle `null`. 

PS. Thanks for the library.

**Comment by ehoner** [comment]
I believe this is unnecessary and not a valuable addition. I should have closed this long ago, I left it open to think about examples where this would be useful and was not able to identify any. Then forgot about the PR, sorry for letting this linger. 

Thanks again for your help @fthomas and the library.

**Pull request closed without merging**